ext/dom/php_dom.c

```diff
@@ -1231,25 +1239,38 @@ xmlNode *dom_get_elements_by_tag_name_ns_raw(xmlNodePtr nodep, char *ns, char *l
 
 	while (nodep != NULL && (*cur <= index || index == -1)) {
 		if (nodep->type == XML_ELEMENT_NODE) {
-			if (xmlStrEqual(nodep->name, (xmlChar *)local) || xmlStrEqual((xmlChar *)"*", (xmlChar *)local)) {
-				if (ns_match_any || (!strcmp(ns, "") && nodep->ns == NULL) || (nodep->ns != NULL && xmlStrEqual(nodep->ns->href, (xmlChar *)ns))) {
+			if (local_match_any || xmlStrEqual(nodep->name, (xmlChar *)local)) {
+				if (ns_match_any || (ns[0] == '\0' && nodep->ns == NULL) || (nodep->ns != NULL && xmlStrEqual(nodep->ns->href, (xmlChar *)ns))) {
 					if (*cur == index) {
 						ret = nodep;
 						break;
 					}
 					(*cur)++;
 				}
 			}
-			ret = dom_get_elements_by_tag_name_ns_raw(nodep->children, ns, local, cur, index);
-			if (ret != NULL) {
-				break;
+
+			if (nodep->children) {
+				nodep = nodep->children;
+				continue;
 			}
 		}
-		nodep = nodep->next;
+
+		if (nodep->next) {
+			nodep = nodep->next;
+		} else {
+			/* Go upwards, until we find a parent node with a next sibling, or until we hit the base. */
+			do {
+				nodep = nodep->parent;
+				ZEND_ASSERT(nodep != NULL);
```

Comment:
It's fine I guess. It just seems like a lot of faith to have in your code and the code of future developers. If I simulate a forgotten cache invalidation, say by commenting out the one in DOMNode::removeChild(), two of the tests crash here. Something like the following suggestion would be safer, would following existing conventions in the file, and would add minimal overhead.

```suggestion
				if (UNEXPECTED(!nodep)) {
					zend_throw_error(NULL, "Current node is not in a document");
					return NULL;
				}
```

Reply:
I would rather have the assertion as this implies a bug within the implementation and having something crash is a very good indicator that we need to fix this.

Having a random Error been thrown to users is not really helpful IMHO, especially as I bet most people will not think to report an issue to the project.

Reply:
There are three options I see right now:
* Assert. This will cause a crash in production which is bad, but it can result in a bug report.
* Just return NULL. This will not result in a crash but causes in wrong results, and might not result in a bug report.
* Throw an error and return NULL. The error might confuse users and might not result in a bug report.

There seems to be no good option here tbh. I like to avoid a crash in production too though... We can write something in the exception like "please report the bug at ...", but this has never been done before in php afaik.

Reply:
We did have such a bug report error for a corrupted timezone database, so it wouldn't be unheard of. So feel free to do something like this.